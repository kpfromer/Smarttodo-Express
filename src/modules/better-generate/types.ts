import { Model, Document } from "mongoose";

export const mongooseTypeMapping = {
  string: String,
  number: Number
}

export type Types = 'string' | 'number' | StructureSchema;
export type CustomRoute = (model: Model<any>) => (req, res, next?) => any;
export type CustomRouteConfig = {
  type: 'get' | 'post' | 'put' | 'patch' | 'delete' | 'all',
  path?: string,
  validation?: 'full' | 'partial' | StructureSchema | null,
  route: CustomRoute
}

export type SingleProp = {
  isArray?: boolean,
  type: Types
  required?: boolean
}

export type StructureSchema = {
  [property: string]: SingleProp
};

export type RedefinedRoute = ((req, res) => any) | boolean;

export type Structure = {
  databaseName: string,
  definedRoutes?: {
    getAll?: RedefinedRoute,
    getSingle?: RedefinedRoute,
    post?: RedefinedRoute,
    put?: RedefinedRoute,
    patch?: RedefinedRoute,
    delete?: RedefinedRoute
  }
  schema: StructureSchema,
  authenticated: boolean,
  routes?: CustomRouteConfig[]
}