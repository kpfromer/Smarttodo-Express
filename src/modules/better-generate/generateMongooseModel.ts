import { mongooseTypeMapping, SingleProp, StructureSchema, Structure } from "./types";
import mongoose, { SchemaTypeOpts, SchemaDefinition, Schema } from "mongoose";

export const mongooseModelSingleProp = ({ type, isArray = false, required = false }: SingleProp): SchemaTypeOpts<any> => {
  let mongooseType;
  if (typeof type === 'object') {
    mongooseType = mongooseModelObject(type);
  } else {
    mongooseType = mongooseTypeMapping[type];
  }
  
  if (isArray) {
    mongooseType = [mongooseType];
  }

  return {
    type: mongooseType,
    required
  }
}

export const mongooseModelObject = (object: StructureSchema): SchemaDefinition => {
  return Object.entries(object).reduce((prev, [propName, schema]) => 
    ({ ...prev, [propName]: mongooseModelSingleProp(schema) }),
  {});
}

export const createMongooseModel = (structure: Structure) => {
  return mongoose.model(structure.databaseName, new mongoose.Schema(mongooseModelObject(structure.schema)));
};