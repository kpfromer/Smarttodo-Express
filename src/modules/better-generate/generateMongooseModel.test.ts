import * as generate from './generateMongooseModel';
import { mongooseModelObject, mongooseModelSingleProp } from "./generateMongooseModel";
import { SingleProp, Structure } from "./types";
import mongoose from 'mongoose';

jest.mock('mongoose', () => ({
  model: jest.fn(),
  Schema: jest.fn()
}))

describe('mongooseModelSingleProp', () => {
  describe('mongoose type', () => {
    it('recursively runs mongooseModelObject if type is an object', () => {
      const schema = {
        type: {
          subItemTitle: { type: 'string' },
          subItemAge: { type: 'number' }
        }
      } as SingleProp;
      expect(mongooseModelSingleProp(schema)).toMatchSnapshot();
    });
    it('sets type to array if isArray', () => {
      const result = mongooseModelSingleProp({ type: 'string', isArray: true});
      expect(result).toMatchSnapshot();
      expect(result.type.length).toBe(1);
      expect(result.type[0]).toBe(String);
    });

    it('maps Number to "number"', () => {
      expect(mongooseModelSingleProp({ type: 'string' })).toMatchObject({
        type: String
      });
    });
    it('maps String to "string"', () => {
      expect(mongooseModelSingleProp({ type: 'number' })).toMatchObject({
        type: Number
      });
    });
  });

  describe('mongoose required', () => {
    it('defaults to false', () => {
      expect(mongooseModelSingleProp({ type: 'string' })).toMatchObject({ required: false });
    });
    it('sets required to user input', () => {
      expect(mongooseModelSingleProp({ type: 'string', required: true })).toMatchObject({ required: true });
    });
  });
});

describe('mongooseModelObject', () => {
  it('converts each key into a proper mongoose option', () => {
    expect(mongooseModelObject({
      name: { type: 'string', required: true },
      project: { type: 'string', required: true }
    })).toMatchSnapshot();
  })
});

describe('createMongooseModel', () => {
  let createMongooseModel, mockModel, mockSchema, mockMongooseModelObject;
  beforeEach(() => {
    mockModel = jest.fn();
    mockSchema = jest.fn();

    jest.doMock('mongoose', () => ({
      model: mockModel,
      Schema: mockSchema
    }));

    // mockMongooseModelObject = jest.fn();
    mockMongooseModelObject = jest.spyOn(generate, 'mongooseModelObject')
    // jest.doMock('./generateMongooseModel', () => ({
    //   ...require.requireActual('./generateMongooseModel'),
    //   mongooseModelObject: mockMongooseModelObject
    // }))

    createMongooseModel = require('./generateMongooseModel').createMongooseModel;
    // console.log({ createMongooseModel });
  });

  it('creates a model using the dynamically generated schema', () => {
    const userStructure = {
      databaseName: 'testitem',
      schema: {
        name: { type: 'string', required: true },
        project: { type: 'string', required: true }
      },
      authenticated: false
    } as Structure;


    (mongoose.model as any).mockImplementation((dbName, schema) => `Model called "${dbName}" of "${schema}"`); // TODO: FIX
    (mongoose.Schema as any).mockImplementation(value => `Schema of ${value}`);
    mockMongooseModelObject.mockReturnValue('generated typings');

    const result = createMongooseModel(userStructure);
    
    expect(result).toMatchSnapshot();
    expect(mockMongooseModelObject).toHaveBeenCalledWith(userStructure.schema)
  })
});