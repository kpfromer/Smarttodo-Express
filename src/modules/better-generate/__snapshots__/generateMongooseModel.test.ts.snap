// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMongooseModel creates a model using the dynamically generated schema 1`] = `"Model called \\"testitem\\" of \\"[object Object]\\""`;

exports[`mongooseModelObject converts each key into a proper mongoose option 1`] = `
Object {
  "name": Object {
    "required": true,
    "type": [Function],
  },
  "project": Object {
    "required": true,
    "type": [Function],
  },
}
`;

exports[`mongooseModelSingleProp mongoose type recursively runs mongooseModelObject if type is an object 1`] = `
Object {
  "required": false,
  "type": Object {
    "subItemAge": Object {
      "required": false,
      "type": [Function],
    },
    "subItemTitle": Object {
      "required": false,
      "type": [Function],
    },
  },
}
`;

exports[`mongooseModelSingleProp mongoose type sets type to array if isArray 1`] = `
Object {
  "required": false,
  "type": Array [
    [Function],
  ],
}
`;
